### PR TITLE
Setup wlr renderer backend

### PR DESCRIPTION
**Pull request by wineee**
forces the creation of a specified renderer if 'WLR_RENDERER' was set

### CODE REVIEW

File: examples/tinywl/main.cpp [hunk 1]
```diff
@@ -483,8 +484,40 @@ OutputInfo* Helper::getOutputInfo(WOutput *output)
...
+    const auto wlrRenderer = qgetenv("WLR_RENDERER");
+
+    if (wlrRenderer == "auto" || wlrRenderer.isEmpty()) {
+        QList<QSGRendererInterface::GraphicsApi> apiList = {
+            QSGRendererInterface::OpenGL,
+            QSGRendererInterface::Software
+        };
+        QQuickWindow::setGraphicsApi(WRenderHelper::probe(apiList));
+    }
+    else if (wlrRenderer == "gles2") {
```
Comment (zccrs) [inline examples/tinywl/main.cpp:503]: style: `} else if (`

File: src/server/qtquick/wrenderhelper.cpp [hunk 2]
```diff
@@ -566,6 +572,68 @@ QWRenderer *WRenderHelper::createRenderer(QWBackend *backend)
...
+
+        if (formats && formats->len == 0) {
+            qInfo() << GraphicsApiName(api) << " api don't support any format";
+            continue;
+        }
+
+        // TODO: how to test when formats gets NULL
+        if (formats && formats->len) {
+            auto *format = &formats->formats[0];
+            auto *alloc = wlr_allocator_autocreate(backend->handle(), renderer->handle());
```
Comment (zccrs) [inline src/server/qtquick/wrenderhelper.cpp:615]: use std::unique_ptr

File: src/server/qtquick/wrenderhelper.cpp [hunk 3]
```diff
@@ -566,6 +572,68 @@ QWRenderer *WRenderHelper::createRenderer(QWBackend *backend)
...
+    default:
+        return "Unknown/Unsupported";
+    }
+}
+
+QSGRendererInterface::GraphicsApi WRenderHelper::probe(const QList<QSGRendererInterface::GraphicsApi> &apiList)
+{
+    auto acceptApi = QSGRendererInterface::Unknown;
+    auto display = new QWDisplay();
+    auto backend = QWBackend::autoCreate(display);
```
Comment (zccrs) [inline src/server/qtquick/wrenderhelper.cpp:594]: Allows provide QWBackend by user

**Review by zccrs — CHANGES_REQUESTED** [review]

File: examples/tinywl/main.cpp [hunk 4]
```diff
@@ -483,8 +484,40 @@ OutputInfo* Helper::getOutputInfo(WOutput *output)
     return infoPtr;
 }
 
+static void setupRendererBackend()
```
Comment (zccrs) [inline examples/tinywl/main.cpp:487]: move to WRenderHelper

**Review by zccrs — CHANGES_REQUESTED** [review]

File: examples/tinywl/main.cpp [hunk 5]
```diff
@@ -483,8 +484,40 @@ OutputInfo* Helper::getOutputInfo(WOutput *output)
...
+{
+    if (qEnvironmentVariableIsSet("QSG_RHI_BACKEND")) {
+        qWarning("Tinywl will ignore `QSG_RHI_BACKEND`, please set `WLR_RENDERER` instead!");
+        qunsetenv("QSG_RHI_BACKEND");
```
Comment (zccrs) [inline examples/tinywl/main.cpp:491]: No, we should allow user to set RHI backend

**Review by zccrs — CHANGES_REQUESTED** [review]

File: src/server/qtquick/wrenderhelper.cpp [hunk 6]
```diff
@@ -566,6 +572,108 @@ QWRenderer *WRenderHelper::createRenderer(QWBackend *backend)
...
+
+        if (formats && formats->len == 0) {
+            qInfo() << GraphicsApiName(api) << " api don't support any format";
+            continue;
+        }
+
+        // TODO: how to test when formats gets NULL
+        if (formats && formats->len) {
+            auto *format = &formats->formats[0];
+            auto *alloc = wlr_allocator_autocreate(testBackend->handle(), renderer->handle()); // destroy follow backend
```
Comment (zccrs) [inline src/server/qtquick/wrenderhelper.cpp:651]: Where is the codes of `destroy follow backend`

File: src/server/qtquick/wrenderhelper.cpp [hunk 7]
```diff
@@ -566,6 +572,108 @@ QWRenderer *WRenderHelper::createRenderer(QWBackend *backend)
...
+{
+    const auto wlrRenderer = qgetenv("WLR_RENDERER");
+
+    if (wlrRenderer == "auto" || wlrRenderer.isEmpty()) {
+        if (qEnvironmentVariableIsSet("QSG_RHI_BACKEND")) {
+            // when environment variable QSG_RHI_BACKEND was set, don't call setGraphicsApi
+            return;
+        }
+
+        QList<QSGRendererInterface::GraphicsApi> apiList = {
```
Comment (zccrs) [inline src/server/qtquick/wrenderhelper.cpp:600]: TODO: Add vulkan to list.

**Review by zccrs — CHANGES_REQUESTED** [review]

File: src/server/qtquick/wrenderhelper.cpp (hunk 6 above)
Comment (wineee) [reply src/server/qtquick/wrenderhelper.cpp:651]: backend_destroy -> finish_drm_renderer -> wlr_allocator_destroy

File: src/server/qtquick/wrenderhelper.cpp [hunk 8]
```diff
@@ -566,6 +572,117 @@ QWRenderer *WRenderHelper::createRenderer(QWBackend *backend)
...
+        if (!testBackend) {
+            display = new QWDisplay();
+            testBackend = QWBackend::autoCreate(display);
+
+            if (!testBackend)
+                qFatal("Failed to create wlr_backend");
+        }
+        QQuickWindow::setGraphicsApi(WRenderHelper::probe(testBackend, apiList));
+        if (display)
+            wl_display_destroy(display->handle());
```
Comment (zccrs) [inline src/server/qtquick/wrenderhelper.cpp:615]: should use `delete display` here, and can use `std::unique_ptr`

File: src/server/qtquick/wrenderhelper.cpp (hunk 6 above)
Comment (wineee) [reply src/server/qtquick/wrenderhelper.cpp]: I can't reproduce the crash，you are right， there are not double free

**Review by zccrs — CHANGES_REQUESTED** [review]

**Review by zccrs — APPROVED** [review]

**Pull request merged**